paperqa/llms.py

```diff
@@ -197,6 +197,69 @@ async def embed_documents(self, texts):
         )
         return np.concatenate(all_embeds, axis=1)
 
+    def set_mode(self, mode: EmbeddingModes) -> None:
+        # Set mode for all component models
+        for model in self.models:
+            model.set_mode(mode)
+
+
+class SentenceTransformerEmbeddingModel(EmbeddingModel):
+    """An embedding model using SentenceTransformers."""
+
+    name: str = Field(default="multi-qa-MiniLM-L6-cos-v1")
+    config: dict[str, Any] = Field(default_factory=dict)
+    _model: Any = None
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        try:
+            from sentence_transformers import SentenceTransformer
+        except ImportError as exc:
+            raise ImportError(
+                "Please install sentence-transformers to use SentenceTransformerEmbeddingModel."
+            ) from exc
+
+        self._model = SentenceTransformer(self.name)
+
+    def set_mode(self, mode: EmbeddingModes) -> None:
+        """Set the embedding mode. SentenceTransformer does not support modes, so this is a no-op."""
+        # SentenceTransformer does not support different modes.
+
+    async def embed_documents(self, texts: list[str]) -> list[list[float]]:
+        """
+        Asynchronously embed a list of documents using SentenceTransformer.
+
+        Args:
+            texts: A list of text documents to embed.
+
+        Returns:
+            A list of embedding vectors.
+        """
+        loop = asyncio.get_running_loop()
+        # Extract additional configurations if needed
+        batch_size = self.config.get("batch_size", 32)
+        device = self.config.get("device", "cpu")
+
+        # Update the model's device if necessary
+        if device:
+            self._model.to(device)
+
+        # Run the synchronous encode method in a thread pool to avoid blocking the event loop.
+        embeddings = await loop.run_in_executor(
```

Comment:
Let's use `asyncio._to_thread` to avoid all nesting troubles associated with `asyncio.get_running_loop`

Reply:
TIL - should we do that everywhere in `docs.py`?

Reply:
Yeah we should, can leave for another PR though